[BE]그룹 생성 페이지 - 그룹 생성 페이지 API

ChanYangYu opened the pull request:
## 관련 이슈

- closes #20

## 셀프 체크리스트

> 최소 1명 이상의 assign을 받아야만 Merge 가능합니다.

- [x] Warning Message 발생 유/무
- [x] Coding Convention 준수 유/무

## 작업 요약
> 작업 내역 요약
- GroupCreatePage 내부 GET요청 로직 구현
- POST 요청 로직 구현
- 기타 코드 개선 : alias hook type 통일 적용
## Preview

> 선택 사항
- Group 테이블 
![image](https://user-images.githubusercontent.com/55623688/140858440-3639bf41-207d-4e6a-badb-599c0d8f366b.png)
- UsingTechStack 테이블
![image](https://user-images.githubusercontent.com/55623688/140858481-72cd25e7-303d-4d16-bd5e-4d9c2f14eb10.png)

## PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
> 백로그에 없지만 필요하여 구현한 기능 

- Group.OwnerId는 아직 몰라서 0으로 저장하도록 했습니다!
- Group.mentoId경우 처음 그룹생성시 없을 수 있기 때문에 Nullable로 변경했습니다.
- Group.Category를 ManyToOne으로 변경했습니다
- UsingTechStack.techStack을 ManyToOne으로 변경했습니다.
- 기본적인 POST 요청할때 발생하는 DB저장 로직은 그룹을 만들고 받아온 usingTechStack 배열의 기술스택을 그룹과 매핑시켜서 저장하도록 동작합니다.

ChanYangYu committed:
[FE] feat: CategoryAPI Response값 적용  


- 기존 dummy데이터 -> Category정보를 getCategory API Response값 적용
- 카테고리 이미지 적용 (imageURL)
- CategoryItem: ClickedItem, UnclickedItem을 Item을 상속받는 styledComponent로 변경

ChanYangYu committed:
[FE] refactor: API호출 로직 통합관리


- category, techStack API호출 로직 GroupCreatePage/index에서 한번만 요청하도록 로직변경
- issue (#20)

ChanYangYu committed:
[FE] feat: Group API 파일 및 타입 생성


- group API 파일 생성 및  axios 추상화 API함수 구현
- POS요청에 필요한 CreateGroupInterface ServerData에 추가
- issue (#20)

ChanYangYu committed:
[BE] feat: Group 도메인 API 구현


- Controller, Service, Repository 파일 생성 및 구현
- dto 디렉토리 생성 및 CreateGroupDto 파일 추가
- issue (#20)

ChanYangYu committed:
[FE] feat: GroupCreate POST 요청 로직 추가


- 마지막 페이지에서 완료 버튼 클릭시 requestGroupCreate 함수 실행 로직 추가 
- requestGroupCreate : Redux전역상태의 저장중인 그룹생성 데이터를 ServerData에 맞게 생성 후 서버에 POST 요청
- Router에 GroupCreatePage 추가
- issue (#20)

ChanYangYu committed:
[FE] rename:  그룹생성 관련 파일 alias 적용


- GroupCreatePage alias 적용
- api Group alias 적용
- issue (#20)

ChanYangYu committed:
[FE] rename: redux hook 적용


- Slice 이름 변경 : createGroupDataSlice ->groupCreateDataSlice\
- Slice State 선언 및 export : groupCreateDataState
- useAppSelector, useAppDispatch 적용

ChanYangYu committed:
[FE] refactor: 타입 단일화


- CreateGroup 타입을 ServerData/GroupCreateInterface 타입으로 통일
  - slice내부 state 타입 GroupCreateInterface로 변경
  - GroupCreatePage 내부 타입 변경
- 타입 단일화에 따른 requestGroupCreate 함수 코드 개선

ChanYangYu committed:
[BE] fix: Domain Model 수정


- Group.mentoId경우 처음 그룹생성시 없을 수 있기 때문에 Nullable로 변경
- Group.Category를 @ManyToOne으로 변경
- UsingTechStack.techStack을 @ManyToOne으로 변경
- TechStack.usingTechStack을 @OneToMany으로 변경

ChanYangYu committed:
[BE] fix: DB Insert 방법 변경


- repository.insert -> repository.save : 리턴값을 저장된 Entity값으로 받기 위해 변경

ChanYangYu committed:
[BE] feat: 그룹 생성 POST 로직 구현


- POST 요청시 그룹 생성(카테고리 매핑) 및 그룹 사용기술스택 생성
- issue (#20)

yuchanleeme commented on server/src/domains/group/service/GroupService.ts:
```diff
@@ -0,0 +1,58 @@
+
+    const createdGroup = await this.groupRepository.createGroup(group);
+
+    groupData.usingTechStacks.forEach(async (techStackName) => {
```
여기 부분은 따로 함수로 분리해도 좋을것 같습니다! usingTechStack의 어떤부분을 처리하는지 함수이름으로 알려줘도 좋을것 같아요!

yuchanleeme replied:
그리고 지금 여기 부분 로직 같은경우는 techstack 쪽에서 service 함수를 만들어서 처리하는건 어떤가요?? 
49번째줄에 `usingTechStack.group = createdGroup;`를 제외하고는 모두 techstack의 모델과 repository를 import해서 사용하는거 같습니다. createdGroup 부분만 인자로 넘겨줘서 사용해도 괜찮을것 같습니다!

ChanYangYu replied:
따로 분리해서 이제  TechStack관련 로직은 모두 TechStackService로 이전 시켰습니다!

ChanYangYu committed:
[BE] refactor: createUsingTechStack 로직 분리


- GroupService에서 TechStackService로 createUsingTechStack분리
- GroupController에 TechStackService 추가
- GroupController.create 함수 async로 변경

Donggggg commented on client/src/pages/GroupCreatePage/TechStackInput/SelectedTechStackList.tsx:
```diff
@@ -1,26 +1,22 @@
     const targetTechStack = e.currentTarget as HTMLButtonElement;
     const nowTechStack = targetTechStack.previousElementSibling?.innerHTML;
-    const newSelectedTechStack = selectedTechStack.filter((techStackName) => {
+    const newSelectedTechStack = usingTechStacks.filter((techStackName) => {
       return techStackName !== nowTechStack;
     });
```
취향차이지만.. 중괄호와 return을 제거하면 1줄로 줄일 수 있을 것 같아요!

ChanYangYu replied:
엇..이게 Ctrl+S로 prettier 적용시키면서 자동으로 내려간거같네요.. 저도 두줄로 되는거 싫어해서 분명 한줄로 작성했던걸로 기억하는데.. 어렵군요..

Donggggg replied:
앗 그건 어쩔수 없는 것 같네요...ㅠㅠ

Donggggg commented on client/src/pages/GroupCreatePage/index.tsx:
```diff
@@ -64,15 +69,38 @@ function GroupCreatePage(): JSX.Element {
 
-  const cleanUp = () =>{
+  const requestGroupCreate = async () => {
+    try {
+      await postGroupCreate(groupData);
+      window.location.href = '/';
+    } catch (e) {
+      console.log(e);
+    }
+  };
```
http 통신 꼼꼼한 예외처리 좋네요!
추후에 실패했을 때 모달도 띄워주면 UX적으로 굉장히 좋을 것 같습니다

ChanYangYu replied:
오호 좋은 생각이네요!! 나중에 모달 컴포넌트가 완성되면 추가하겠습니다!

Donggggg commented on client/src/pages/GroupCreatePage/index.tsx:
```diff
@@ -64,15 +69,38 @@ function GroupCreatePage(): JSX.Element {
+    fetchCategoryList();
+    fetechTechStackList();
+
     return () => cleanUp();
```
cleanUp()함수를 분리해서 의미가 분명하네요..!
길어지면 저도 사용해보겠습니다!

ChanYangYu replied:
ㅎㅎ감사합니다~

Donggggg commented on server/src/domains/group/service/GroupService.ts:
```diff
@@ -0,0 +1,39 @@
+    group.intro = groupData.intro;
+    group.startAt = new Date(groupData.startAt);
+    group.endAt = new Date(groupData.endAt);
+
+    const createdGroup = await this.groupRepository.createGroup(group);
+
+    return createdGroup;
+  }
```
클라이언트 측에서는 해당 api의 응답 body를 사용하지 않던데 body를 같이 전송하신 이유가 있을까요!?

Donggggg replied:
코드를 잘못 링킹했네요.
언급한 코드가 아닌 GroupController부분으로 생각하시면 될 것 같아요!

ChanYangYu replied:
아하 이부분은 리턴을 무슨값 하지않으면 오류가 나서 Group을 리턴하도록 했습니다. 혹시 여기에 무슨정보를 리턴해야하는게 맞을까요?

Donggggg replied:
제가 사용한 OnUndefined 데코레이터 참고하시면 좋을 것 같아요.
리턴값이 없으면 OnUndefined(statusCode)에 있는 statusCode의 응답을 보내줘요!

yuchanleeme reviewed (approved):
코드 짜시느라 수고많으셨습니다!! 👍👍

ChanYangYu committed:
[BE] style: api 리턴값 OnUndefined적용 


- api 리턴값 OnUndefined적용
- issue(#20)

Donggggg reviewed (approved):
수정 사항 확인했습니다!
고생하셨어용 👍

ChanYangYu committed:
[BE] fix: rebase conflict 처리완료


- rebase 이후 에러 해결
- 누락코드 추가 (확인필요)
- createGroupUsingStack UsingTechStackService로 이동

who-is-hu commented on server/src/domains/group/service/GroupService.ts:
```diff
@@ -1,16 +1,42 @@
+    group.intro = groupData.intro;
+    group.startAt = new Date(groupData.startAt);
+    group.endAt = new Date(groupData.endAt);
+
```
dto->entity를 만드는 과정같은데
이런 코드는 dto에 함수로 작성하는게 좋습니다.
```
class Dto {
  toEntity(){
    ...
   return new Entity();
 }
}
```
이런 식으로요 재사용도 가능하고 분리도됩니다.
